xdsl/transforms/stencil_to_csl_stencil.py

```diff
@@ -0,0 +1,173 @@
+    csl_stencil.access operates on memref< num_neighbors x tensor< buf_size x data_type >> found at last arg index
+
+    Note: This is intended to be called in a nested pattern rewriter, such that the above precondition is met.
+    """
+
+    arg_index: int
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: stencil.AccessOp, rewriter: PatternRewriter, /):
+        assert len(op.offset) == 2
+        if tuple(op.offset) == (0, 0):
+            return
+        prefetched_arg = op.get_apply().region.block.args[-1]
+        assert isa(m_type := prefetched_arg.type, memref.MemRefType[Attribute])
+        assert isa(t_type := m_type.get_element_type(), TensorType[Attribute])
+
+        csl_access_op = csl_stencil.AccessOp(
+            op=prefetched_arg,
+            offset=op.offset,
+            offset_mapping=op.offset_mapping,
+            result_type=t_type,
+        )
+
+        # The stencil-tensorize-z-dimension pass inserts tensor.ExtractSliceOps after stencil.access to remove ghost cells.
+        # Since ghost cells are not prefetched, these ops can be removed again. Check if the ExtractSliceOp
+        # has no other effect and if so, remove both.
+        if (
+            len(op.res.uses) == 1
+            and isinstance(use := list(op.res.uses)[0].operation, tensor.ExtractSliceOp)
+            and tuple(d.data for d in use.static_sizes.data) == t_type.get_shape()
+            and tuple(d.data for d in use.static_offsets.data) == (0,)
+            and tuple(d.data for d in use.static_strides.data) == (1,)
+            and len(use.offsets) == 0
+            and len(use.sizes) == 0
+            and len(use.strides) == 0
+        ):
+            rewriter.replace_op(use, csl_access_op)
+            rewriter.erase_op(op)
+        else:
+            rewriter.replace_matched_op(csl_access_op)
+
+
+@dataclass(frozen=True)
+class SwapToPrefetchPattern(RewritePattern):
+    """
+    Translates dmp.swap to csl_stencil.prefetch
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: dmp.SwapOp, rewriter: PatternRewriter, /):
+        # remove op if it contains no swaps
+        if op.swaps is None or len(op.swaps) == 0:
+            rewriter.erase_matched_op(False)
+            return
+        assert isinstance(op.input_stencil, OpResult)
+
+        # currently only works for 3-dimensional stencils
+        assert all(len(swap.size) == 3 for swap in op.swaps)
+
+        # dmp should decompose from (x,y,z) to (1,1,z)
+        assert all(swap.size[:2] == (1, 1) for swap in op.swaps)
+
+        # check that size is uniform
+        uniform_size = op.swaps.data[0].size[2]
+        assert all(swap.size[2] == uniform_size for swap in op.swaps)
+
+        assert isa(
+            op.input_stencil.type,
+            memref.MemRefType[Attribute] | stencil.TempType[Attribute],
+        )
+        assert isa(
+            t_type := op.input_stencil.type.get_element_type(), TensorType[Attribute]
+        )
```

Comment:
Still no messages on these asserts

Reply:
Added assert messages for non-type check asserts